scylla/src/transport/connection.rs

```diff
@@ -820,6 +822,59 @@ impl Connection {
         }
     }
 
+    async fn prepare_batch<'b>(
+        &self,
+        init_batch: &'b Batch,
+        values: impl BatchValues,
+    ) -> Result<Cow<'b, Batch>, QueryError> {
+        let mut to_prepare = HashSet::<String>::new();
+
+        {
+            let mut values_iter = values.batch_values_iter();
+            for stmt in &init_batch.statements {
+                if let BatchStatement::Query(query) = stmt {
+                    let value = values_iter.next_serialized().transpose()?;
+                    if let Some(v) = value {
+                        if v.len() > 0 {
+                            to_prepare.insert(query.contents.clone());
+                        }
+                    }
+                } else {
+                    values_iter.skip_next();
+                }
+            }
+        }
+
+        let mut prepared_queries = HashMap::<String, PreparedStatement>::new();
+
+        for query in &to_prepare {
+            let prepared = self.prepare(&Query::new(query)).await?;
+            prepared_queries.insert(query.clone(), prepared);
+        }
+
+        let mut batch: Cow<Batch>;
+
+        if to_prepare.is_empty() {
+            batch = Cow::Borrowed(init_batch);
+        } else {
+            batch = Cow::Owned(Default::default());
+            batch.to_mut().config = init_batch.config.clone();
+            for stmt in &init_batch.statements {
+                match stmt {
+                    BatchStatement::Query(query) => match prepared_queries.get(&query.contents) {
+                        Some(prepared) => batch.to_mut().append_statement(prepared.clone()),
+                        None => batch.to_mut().append_statement(query.clone()),
+                    },
+                    BatchStatement::PreparedStatement(prepared) => {
+                        batch.to_mut().append_statement(prepared.clone());
+                    }
+                }
+            }
+        }
```

Comment:
In Rust, `if`s are expressions and evaluate to a value - you can use that to assign to `batch` immediately.

Moreover, in the `else` part, I suggest creating a value of type `Batch`, initializing it and then assigning it to `Cow<Batch>`. This will allow to avoid calling `batch.to_mut()` in multiple places.

```rust
let batch = if to_prepare.is_empty() {
    Cow::Borrowed(init_batch)
} else {
    let mut batch = Batch::default();
    // ...
    Cow::Owned(batch)
}
```

Reply:
Though, if you implement an early return like I suggested before, here you will be able to drop the `if` altogether and only leave the code from the `else` branch.